### PR TITLE
Improve GPU detection

### PR DESCRIPTION
**Pull request by vchuravy**
This is a quality of life improvement to our libmxnet build process. I have tested this on my local Linux box, but I would appreciate wider testing (especially on Windows)

Fixes: #175, #143

### CODE REVIEW

**Comment by tkelman** [comment]
Whoa, whoa, whoa. What the hell is `usr/setupenv.cmd` doing?
```
REM @echo off

setlocal EnableExtensions EnableDelayedExpansion

REM Remove duplicate entries in PATH to avoid long PATH.
set _PATH_=
for %%a in ("%PATH:;=" "%") do @if not "%%~a" == "" (
if "!_PATH_!" == "" @set "_PATH_=;%%~a;"
set "_T_=!_PATH_:;%%~a;=x!"
if "!_T_!" == "!_PATH_!" @set "_PATH_=!_PATH_!%%~a;"
)
set "_path=%_PATH_:~1,-1%
set _path=%_path%;

if [%MXNET_HOME%] == [] GOTO Define
    echo Deleting original install
    set _root=%MXNET_HOME%
    set _dirs=(%_root%\3rdparty\openblas\bin %_root%\3rdparty\gnuwin %_root%\3rdparty\vc %_root%\3rdparty\cudart %_root%\3rdparty\cudnn\bin %_root%\lib)
    for %%d in %_dirs% do set _path=!_path:%%d;=!
    echo PATH=%_path%

:Define
    setlocal EnableExtensions EnableDelayedExpansion
    echo Defining new environmental variable...
    set _root=%~dp0
    setx MXNET_HOME %_root%
    set _root=%%MXNET_HOME%%
    set _dirs=(%_root%\3rdparty\openblas\bin %_root%\3rdparty\gnuwin %_root%\3rdparty\vc %_root%\3rdparty\cudart %_root%\3rdparty\cudnn\bin %_root%\lib)
    for %%d in %_dirs% do set _path=%%d;!_path!
    setx PATH "%_path%"
    echo MXNET_HOME=%~dp0
    echo PATH=%_path%
```

`setx` is an absolute no-no. Adding things permanently to the path is a recipe for breaking anything else that depends on a dll with the same name. Shut that off NOW. The Julia package can modify path, locally, temporarily, in `__init__` if it needs to. Not globally outside of Julia.

**Comment by vchuravy** [comment]
Will do and raise that issue with the main MXNet group. Cc @pluskid

**Comment by tkelman** [comment]
I'd recommend commenting that line out and issuing a replacement release ASAP, permanent environment variable modifications is not something I want users to hit.

**Comment by tkelman** [comment]
On a less "angry person yelling at you" note, my system apparently has nvcuda.dll easily findable and loadable (nvidia put it in system32 at some point? grumble grumble...), but not a bunch of other cu* or nv* dll's. I don't remember ever installing cuda on my Windows partition - I have a GTX 960M but I haven't done much with it lately.

image

The error message from trying to load MXNet.jl was `Cannot find or load libmxnet.so. Please see the document on how to build it.` which isn't the greatest suggestion for people on Windows. The call to `Libdl.find_library(["libmxnet.so","libmxnet.dll"], [joinpath("$(get(ENV,"MXNET_HOME",""))","lib"), joinpath(Pkg.dir("MXNet"),"deps/usr/lib")])` was failing, and speaking of not-very-useful error messages,
```
julia> Libdl.dlopen(Pkg.dir("MXNet","deps","usr","lib","libmxnet.dll"))
ERROR: could not load library "C:\Users\Tony\.julia\v0.5\MXNet\deps\usr\lib\libmxnet.dll"
The operation completed successfully.

 in dlopen(::String, ::UInt32) at .\libdl.jl:90 (repeats 2 times)
```
which I guess I'll blame on Jameson.

**Comment by vchuravy** [comment]
@musm Can you try this on windows?

**Comment by musm** [comment]
@vchuravy  ok I tried this for you 

One thing is that the link is wrong
It needs the following patch
```
    name = "mxnet_x64_vc14_$(HAS_CUDA ? "gpu" : "cpu").7z"
    package_url = "https://github.com/yajiedesign/mxnet/releases/download/$(curr_win)/$(curr_win)_$(name)"

    exe7z = joinpath(JULIA_HOME, "7z.exe")

    run(download_cmd(base_url, "mxnet_base.7z"))
    run(`$exe7z x mxnet_base.7z -y -ousr`)
    run(`cmd /c copy "usr\\3rdparty\\openblas\\bin\\*.dll" "usr\\lib"`)

    run(download_cmd(package_url, "mxnet.7z"))
    run(`$exe7z x mxnet.7z -y -ousr`)
```

**Comment by musm** [comment]
Another thing is that
```
const MXNET_LIB = Libdl.find_library(["libmxnet.dll"],
                                     [joinpath("$(get(ENV,"MXNET_HOME",""))","lib"),
                                      Pkg.dir("MXNet","deps","usr","lib")])
```
returns empty even though the dll is actually in the folder??? very strange

**Comment by vchuravy** [comment]
Why do we need to use `vc14` vs `vc12`? and not finding the library is very strange indeed. What happens if you try to dlopen the library directly?

**Comment by musm** [comment]
well there is *no* link to *vc12*  i.e. the link you have doesn't exist. I looked at the repo and there is only *vc14*.

**Comment by musm** [comment]
image

**Comment by vchuravy** [comment]
Ok update the base package to also have vc14 and updated the version we try to download.

**Comment by musm** [comment]
oh  ok so I just check this out again, and I guess with the new links  `libmxnet.dll` is actually in the *build* directory and not the *lib* directory
However it is still not possible to open the dll
image


I'm thinking there is something wrong with the dll...

**Comment by pluskid** [comment]
Sometimes the error messages from Julia FFI is not very useful. Maybe trying to use python to load the dll and see if it could load correctly? https://docs.python.org/2/library/ctypes.html#ctypes.LibraryLoader.LoadLibrary

File: deps/build.jl [hunk 1]
```diff
@@ -20,32 +20,66 @@ if haskey(ENV, "MXNET_HOME")
...
     exe7z = joinpath(JULIA_HOME, "7z.exe")
 
     run(download_cmd(base_url, "mxnet_base.7z"))
     run(`$exe7z x mxnet_base.7z -y -ousr`)
     run(`cmd /c copy "usr\\3rdparty\\openblas\\bin\\*.dll" "usr\\lib"`)
+    run(`cmd /c copy "usr\\3rdparty\\opencv\\*.dll" "usr\\lib"`)
 
     run(download_cmd(package_url, "mxnet.7z"))
     run(`$exe7z x mxnet.7z -y -ousr`)
+    run(`cmd /c copy "build//*.dll" "usr\\lib"`)
```
Comment (musm) [inline deps/build.jl]: do you mean `build\\*.dll`

Comment (vchuravy) [inline deps/build.jl]: Yup, no coffee...

**Comment by vchuravy** [comment]
The prebuilt binaries are only available for 64bit

**Pull request merged**